synapse/rest/client/v2_alpha/account.py

```diff
@@ -201,6 +249,50 @@ def on_POST(self, request):
         defer.returnValue((200, ret))
 
 
+class MsisdnThreepidRequestTokenRestServlet(RestServlet):
+    PATTERNS = client_v2_patterns("/account/3pid/msisdn/requestToken$")
+
+    def __init__(self, hs):
+        self.hs = hs
+        super(MsisdnThreepidRequestTokenRestServlet, self).__init__()
+        self.identity_handler = hs.get_handlers().identity_handler
+
+    @defer.inlineCallbacks
+    def on_POST(self, request):
+        body = parse_json_object_from_request(request)
+
+        required = [
+            'id_server', 'client_secret',
+            'country', 'phone_number', 'send_attempt',
+        ]
+        absent = []
+        for k in required:
+            if k not in body:
+                absent.append(k)
+
+        if absent:
+            raise SynapseError(400, "Missing params: %r" % absent, Codes.MISSING_PARAM)
+
+        phoneNumber = None
+        try:
+            phoneNumber = phonenumbers.parse(body['phone_number'], body['country'])
+        except phonenumbers.NumberParseException:
+            raise SynapseError(400, "Unable to parse phone number")
+        msisdn = phonenumbers.format_number(
+            phoneNumber, phonenumbers.PhoneNumberFormat.E164
+        )[1:]
```

Comment:
Yeah, this would be much nicer to read if factored out a bit:

```python
def on_POST(self, request):
    body = parse_json_object_from_request(request)
    check_required_keys(request, ('id_server', 'client_secret', ..))

    msisdn = format_msisdn(body['phone_number'], body['country'])

    existingUid = yield self.store.get_user_id_by_threepid('msisdn', msisdn)
    if existingUid:
        raise

    ret = yield self.identity_handler.requestEmailToken(**body)
    defer.returnValue((200, ret))
```